src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -140,6 +142,32 @@ PartitionTableNoLock(Oid relationId)
 }
 
 
+/*
+ * try_relation_open_nolock opens a relation with given relationId without
+ * acquiring locks. PostgreSQL's try_relation_open() asserts that caller
+ * has already acquired a lock on the relation, which we don't always do.
+ */
+static Relation
+try_relation_open_nolock(Oid relationId)
+{
+	Relation relation = NULL;
+	if (!SearchSysCacheExists1(RELOID, ObjectIdGetDatum(relationId)))
+	{
+		return NULL;
+	}
+
+	relation = RelationIdGetRelation(relationId);
+	if (!RelationIsValid(relation))
```

Comment:
* NB: caller should already have at least AccessShareLock on the
 * relation ID, else there are nasty race conditions.
 
`RelationIdGetRelation` has the above comment. However, Citus was not doing it properly for the last few years anyway as @pykello convinced me on a private chat.